### ISSUE TITLE
The artificial triggering of 'onpopstate' in pushState() causes losing of steps

### ISSUE THREAD

**Issue by martin-g** [browserstate/history.js#47]
Hi,

Initially I posted my problem first at http://getsatisfaction.com/balupton/topics/history_js_back_button_goes_two_states_steps_back.
After some research I found the the problem is caused by the manually triggered 'onpopstate' event in the end of the body of History.pushState(). Triggering the event consumes the state too early and later when the user clicks browser's back button there is no notification for the last registered event.
I.e. there is a single link clicking on which pushes a new state. Clicking three times on this link and then clicking browser's back button will call twice my 'statechange' handler - once for the second click and once for the first one, but none for the third.

**Comment by TomK** [comment]
@cervengoc thanks for your example, i've implemented a similar method.

@dherran I'm not sure if i'm reading your PR correctly, on back/forward you are setting isInternal to true, and when triggered via pushState/replaceState it is false?  I figure that should be the other way around?  In my interpretation Internal would mean "from my internal code".

**Comment by mdingena** [comment]
Hi. I'm new to GitHub, but I signed up for an account just to comment with my use case. I'm using https://github.com/browserstate/history.js/pull/401/files with success. But one thing that bugged me was that the very State which is created on the landing page does not contain the internal flag.

So I added a line in the `storeState` function:

    // Set State.internal
    newState.internal = ( typeof newState.internal === 'undefined' ? true : newState.internal );

or minified:

    h.storeState=function(e){e.internal=(typeof e.internal==='undefined'?1:e.internal);return h.urlToId[e.url]=e.id,h.storedStates.push(h.cloneObject(e)),e}

**Comment by nevace** [comment]
Thanks @dherran and also @mdingena for your fixes. This is working how I expected to now and is a great library. How do we get this merged as @balupton isn't maintaining this anymore?

**Comment by balupton** [comment]
I'm happy to add new owners and maintainers to the project.

**Comment by cervengoc** [comment]
I agree with @TomK, I also think that didactically the word "internal" should refer to programmatically changing state, and using Back/Forward button should be the "non-internal". At least I would expect it this way.

**Comment by mdingena** [comment]
It's probably `internal` because the state change is triggered from the browser itself, rather than - say - clicking a link on a page.

**Comment by cervengoc** [comment]
Yes, that can be a reason. However, in most programming environment "internal" means rather something like I wrote. Maybe would be better to call it something like `clientChange`, which is more self explanatory IMHO.

**Comment by nevace** [comment]
@balupton I'm happy to help maintain this along with others. I know the amount of time you were spending got too much so if any others would also help then that would be great.

**Comment by TomK** [comment]
@cervengoc has a point, remove the problem altogether by calling it something different. Perhaps `isNavigation` ?

**Comment by cervengoc** [comment]
That's why I thought of `isClientChange`, because that covers everything that the "client" does, not the "developer".

**Comment by balupton** [comment]
@nevace sweet will add you, then you can add others as it goes - the main issue is the tests, the tests require manual runs - for two reasons:

1. older versions of safari on iOS (not relevant anymore, as no one uses them) - would have different behaviour of a programatic state change versus an actual back button click by a human
2. browsers sometimes forget and persist state data when you do a few state changes, then navigate to a different website, then hit the back button

So, for each version, I was spending a day or more testing a huge array of browsers manually - if that can be automated using better tooling, then history.js would be easily maintained, however such better tooling is not something I have any experience with.

We have a slack community at https://slack.bevry.me that I can add a history.js room for if maintenance is going to resume.

**Comment by balupton** [comment]
@nevace invite sent

**Comment by balupton** [comment]
Anyone else want to help maintain too?

**Comment by balupton** [comment]
Moving maintenance discussion over to https://github.com/browserstate/history.js/issues/450

**Comment by picard102** [comment]
Where in the file would I insert 

```
// Set State.internal
newState.internal = ( typeof newState.internal === 'undefined' ? true : newState.internal );
```
as per @mdingena's suggestion? The forward button motion seems to work, but any click on the back button results in a false flag for internal.

**Comment by mdingena** [comment]
In the `storeState` function, as I suggested.

**Comment by picard102** [comment]
I tried adding it there, line 1104 in @dherran's edit, still didn't get a true flag on internal when using the back button.  http://dev.podcamptoronto.com/sessions/

**Comment by nevace** [comment]
I'm using this which I've modified with the above and works as expected regarding back/forward.

[jquery.history.js.zip](https://github.com/browserstate/history.js/files/63316/jquery.history.js.zip)

I'm not sure what the best way is to get it to work on first load, I just ended up manually calling History.pushState() on first load. Not very nice to have duplicate entries but works for back/forward (I need to re-init animations etc. on different pages so need to know when back/forward is pressed).

**Comment by danger89** [comment]
Why is this issue still not merged into master??

**Comment by Sitronik** [comment]
@seb78 Thank you, you very helped me.
